app/models/set_member_subject.rb

```diff
@@ -22,47 +22,57 @@ class SetMemberSubject < ActiveRecord::Base
   before_create :set_random
 
   def self.by_subject_workflow(subject_id, workflow_id)
-    joins(subject_set: :workflows)
-      .where(subject_id: subject_id, workflows: {id: workflow_id })
+    by_workflow(workflow_id).where(subject_id: subject_id)
   end
 
-  def self.by_workflow(workflow)
-    joins(:workflows).where(workflows: {id: workflow.id})
-  end
+  def self.by_workflow(workflow_id)
+    linked_workflow_set_ids = SubjectSetsWorkflow
+      .where(workflow_id: workflow_id)
+      .select(:subject_set_id)
 
-  def self.non_retired_for_workflow(workflow)
-    by_workflow(workflow)
-    .joins(sanitize_sql_for_conditions(["LEFT OUTER JOIN subject_workflow_counts ON subject_workflow_counts.subject_id = set_member_subjects.subject_id AND subject_workflow_counts.workflow_id = ?", workflow.id]))
-    .where('subject_workflow_counts.retired_at IS NULL')
+    where(subject_set_id: linked_workflow_set_ids)
   end
 
-  def self.retired_for_workflow(workflow)
-    by_workflow(workflow)
-    .joins(sanitize_sql_for_conditions(["INNER JOIN subject_workflow_counts ON subject_workflow_counts.subject_id = set_member_subjects.subject_id AND subject_workflow_counts.workflow_id = ?", workflow.id]))
-    .where("subject_workflow_counts.retired_at IS NOT NULL")
+  def self.non_retired_for_workflow(workflow_id)
+    non_retired_subject_ids = SubjectWorkflowStatus
+      .where(workflow_id: workflow_id)
+      .where(retired_at: nil)
+      .select(:subject_id)
+
+    by_workflow(workflow_id).where(subject_id: non_retired_subject_ids)
   end
 
-  def self.seen_for_user_by_workflow(user, workflow)
-    seen_subjects = for_user_by_workflow_scope(user, workflow)
-    by_workflow(workflow).where(seen_subjects.exists)
+  def self.retired_for_workflow(workflow_id)
+    retired_subject_ids = SubjectWorkflowStatus
+      .where(workflow_id: workflow_id)
+      .where.not(retired_at: nil)
+      .select(:subject_id)
+
+    by_workflow(workflow_id).where(subject_id: retired_subject_ids)
   end
 
-  def self.unseen_for_user_by_workflow(user, workflow)
-    seen_subjects = for_user_by_workflow_scope(user, workflow)
-    by_workflow(workflow).where(seen_subjects.exists.not)
+  # Be careful using this query as it's not selective on a large table
+  # and the LEFT OUTER JOIN can take a long time to resolve
+  #
+  # Note: push this into where it's being used to ensure it's a more selective
```

Comment:
I'm 50/50 on moving this code. It's only used in 1 place `select_non_retired_unseen_for_user` but it can be a method (costly query to run) if there is a lot of sms data for the workflow. 

Keeping it as a class method provides here seems a nicer interface and this class if responsible for it's search methods. let me know if you want me to move it.

Reply:
I'm not too worried about it if we leave the warning in the comment.